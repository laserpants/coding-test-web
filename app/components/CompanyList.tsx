'use client';

import { useState, useEffect } from 'react';
import type { ICompany } from './company.types';
import Company from './Company';
import LoadingIndicator from './LoadingIndicator';

export default function CompanyList() {
  const [companies, setCompanies] = useState<ICompany[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchCompanies = async () => {
      try {
        setIsLoading(true);
        const response = await fetch('/api/companies');
        if (!response.ok) {
          throw new Error('Failed to fetch companies.');
        }
        const { data: companies } = await response.json();
        setCompanies(companies);
        setError(null);
      } catch (err) {
        setError((err as Error).message);
      } finally {
        setIsLoading(false);
      }
    };

    fetchCompanies();
  }, []);

  if (isLoading) {
    return (
      <LoadingIndicator />
    );
  }

  if (error) {
    return (
      <>
        Error
      </>
    );
  }

  return (
    <div>
      {companies.map((company) => 
        <Company key={company.companyId} {...company} />
      )}
    </div>
  );
}
