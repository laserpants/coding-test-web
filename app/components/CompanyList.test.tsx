import { render, screen } from "@testing-library/react";
import { beforeEach, describe, it, expect, jest } from "@jest/globals";
import { act } from "react";
import CompanyList from "./CompanyList";

global.fetch = jest.fn() as jest.MockedFunction<typeof fetch>;

describe("CompanyList", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", async () => {
    const { container } = await act(async () => render(<CompanyList />));

    expect(container).toMatchSnapshot();
  });

  it("renders the list of companies", async () => {
    (global.fetch as jest.MockedFunction<typeof fetch>).mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        data: [
          { companyId: 1, companyName: "Company A", companyCountry: "US" },
          { companyId: 2, companyName: "Company B", companyCountry: "UK" },
          { companyId: 3, companyName: "Company C", companyCountry: "FI" },
        ],
      }),
    } as Response);

    await act(async () => render(<CompanyList />));

    const listItems = screen.getAllByRole("listitem");

    expect(listItems).toHaveLength(3);
  });

  it("handles fetch failure gracefully", async () => {
    (global.fetch as jest.MockedFunction<typeof fetch>).mockImplementationOnce(
      () => Promise.reject(new Error("Failed to fetch"))
    );

    await act(async () => render(<CompanyList />));

    expect(screen.getByText(/error/i)).toBeInTheDocument();
  });
});
