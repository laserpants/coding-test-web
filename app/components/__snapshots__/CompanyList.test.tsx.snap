// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyList renders correctly 1`] = `
<div>
  <div
    class="bg-red-100 border border-red-400 text-red-700 px-4 py-3 mt-4 rounded relative flex items-center justify-between"
  >
    <div>
      <strong
        class="font-bold"
      >
        Error: 
      </strong>
      <span>
        Cannot read properties of undefined (reading 'ok')
      </span>
    </div>
    <button
      class="bg-red-600 text-white px-4 py-1 rounded hover:bg-red-700 transition"
    >
      Retry
    </button>
  </div>
</div>
`;
